release/src/router/tcpreplay/src/tcprewrite.1
.de1 NOP
.  it 1 an-trap
.  if \\n[.$] \,\\$*\/
..
.ie t \
.ds B-Font [CB]
.ds I-Font [CI]
.ds R-Font [CR]
.el \
.ds B-Font B
.ds I-Font I
.ds R-Font R
.TH tcprewrite 1 "11 Jun 2023" "tcprewrite" "User Commands"
.\"
.\" DO NOT EDIT THIS FILE (in-mem file)
.\"
.\" It has been AutoGen-ed
.\" From the definitions ../../src/tcprewrite_opts.def
.\" and the template file agman-cmd.tpl
.SH NAME
\f\*[B-Font]tcprewrite\fP
\- Rewrite the packets in a pcap file.
.SH SYNOPSIS
\f\*[B-Font]tcprewrite\fP
.\" Mixture of short (flag) options and long options
[\f\*[B-Font]\-flags\f[]]
[\f\*[B-Font]\-flag\f[] [\f\*[I-Font]value\f[]]]
[\f\*[B-Font]\-\-option-name\f[][[=| ]\f\*[I-Font]value\f[]]]
.sp \n(Ppu
.ne 2

All arguments must be options.
.sp \n(Ppu
.ne 2

.SH "DESCRIPTION"
Tcprewrite is a tool to rewrite packets stored in \fIpcap(3)\fP file format,
such as created by tools such as \fItcpdump(1)\fP and \fIwireshark(1)\fP.
Once a pcap file has had it's packets rewritten, they can be replayed back
out on the network using \fItcpreplay(1)\fP.
.sp
tcprewrite currently supports reading the following DLT types:
.sp 1
\fBDLT_C_HDLC\fP aka Cisco HDLC
.sp 1
\fBDLT_EN10MB\fP aka Ethernet
.sp 1
\fBDLT_LINUX_SLL\fP aka Linux Cooked Socket
.sp 1
\fBDLT_RAW\fP aka RAW IP
.sp 1
\fBDLT_NULL\fP aka BSD Loopback
.sp 1
\fBDLT_LOOP\fP aka OpenBSD Loopback
.sp 1
\fBDLT_IEEE802_11\fP aka 802.11a/b/g
.sp 1
\fBDLT_IEEE802_11_RADIO\fP aka 802.11a/b/g with Radiotap headers
.sp 1
\fBDLT_JUNIPER_ETHER\fP aka Juniper Encapsulated Ethernet
.sp 1
\fBDLT_PPP_SERIAL\fP aka PPP over Serial
.sp
Please see the \--dlt option for supported DLT types for writing.
.sp
The packet editing features of tcprewrite which distinguish between "client"
and "server" traffic requires a tcpprep(1) cache file.
.sp
For more details, please see the Tcpreplay Manual at:
http://tcpreplay.appneta.com
.SH "OPTIONS"
.SS ""
.TP
.NOP \f\*[B-Font]\-r\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-portmap\f[]=\f\*[I-Font]string\f[]
Rewrite TCP/UDP ports.
This option may appear up to 9999 times.
.sp
Specify a list of comma delimited port mappings consisting of
colon delimited port number pairs.  Each colon delimited port pair
consists of the port to match followed by the port number to rewrite.
.sp
Examples:
.nf
    \--portmap=80:8000 \--portmap=8080:80    # 80->8000 and 8080->80
    \--portmap=8000,8080,88888:80           # 3 different ports become 80
    \--portmap=8000-8999:80                 # ports 8000 to 8999 become 80
.fi
.TP
.NOP \f\*[B-Font]\-s\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-seed\f[]=\f\*[I-Font]number\f[]
Randomize src/dst IPv4/v6 addresses w/ given seed.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
fuzz-seed.
This option takes an integer number as its argument.
.sp
Causes the source and destination IPv4/v6 addresses to be pseudo
randomized but still maintain client/server relationships.
Since the randomization is deterministic based on the seed,
you can reuse the same seed value to recreate the traffic.
.TP
.NOP \f\*[B-Font]\-N\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-pnat\f[]=\f\*[I-Font]string\f[]
Rewrite IPv4/v6 addresses using pseudo-NAT.
This option may appear up to 2 times.
This option must not appear in combination with any of the following options:
srcipmap.
.sp
Takes a comma delimited series of colon delimited CIDR
netblock pairs.  Each netblock pair is evaluated in order against
the IP addresses.  If the IP address in the packet matches the
first netblock, it is rewritten using the second netblock as a
mask against the high order bits.
.sp
IPv4 Example:
.nf
    \--pnat=192.168.0.0/16:10.77.0.0/16,172.16.0.0/12:10.1.0.0/24
.fi
IPv6 Example:
.nf
    \--pnat=[2001:db8::/32]:[dead::/16],[2001:db8::/32]:[::ffff:0:0/96]
.fi
.TP
.NOP \f\*[B-Font]\-S\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-srcipmap\f[]=\f\*[I-Font]string\f[]
Rewrite source IPv4/v6 addresses using pseudo-NAT.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
pnat.
.sp
Works just like the \--pnat option, but only affects the source IP
addresses in the IPv4/v6 header.
.TP
.NOP \f\*[B-Font]\-D\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-dstipmap\f[]=\f\*[I-Font]string\f[]
Rewrite destination IPv4/v6 addresses using pseudo-NAT.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
pnat.
.sp
Works just like the \--pnat option, but only affects the destination IP
addresses in the IPv4/v6 header.
.TP
.NOP \f\*[B-Font]\-e\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-endpoints\f[]=\f\*[I-Font]string\f[]
Rewrite IP addresses to be between two endpoints.
This option may appear up to 1 times.
This option must appear in combination with the following options:
cachefile.
.sp
Takes a pair of colon delimited IPv4/v6 addresses which will be used to rewrite
all traffic to appear to be between the two IP addresses.
.sp
IPv4 Example:
.nf
    \--endpoints=172.16.0.1:172.16.0.2
.fi
IPv6 Example:
.nf
    \--endpoints=[2001:db8::dead:beef]:[::ffff:0:0:ac:f:0:2]
.fi
.TP
.NOP \f\*[B-Font]\-\-tcp\-sequence\f[]=\f\*[I-Font]number\f[]
Change TCP Sequence (and ACK) numbers /w given seed.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 1
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
Change all TCP sequence numbers, and related sequence-acknowledgement numbers.
They will be shifted by a random amount based on the provided seed.
.TP
.NOP \f\*[B-Font]\-b\f[], \f\*[B-Font]\-\-skipbroadcast\f[]
Skip rewriting broadcast/multicast IPv4/v6 addresses.
.sp
By default \--seed, \--pnat and \--endpoints will rewrite
broadcast and multicast IPv4/v6 and MAC addresses.	Setting this flag
will keep broadcast/multicast IPv4/v6 and MAC addresses from being rewritten.
.TP
.NOP \f\*[B-Font]\-C\f[], \f\*[B-Font]\-\-fixcsum\f[]
Force recalculation of IPv4/TCP/UDP header checksums.
.sp
Causes each IPv4/v6 packet to have their checksums recalculated and
fixed.  Automatically enabled for packets modified with \fB--seed\fP,
\fB--pnat\fP, \fB--endpoints\fP or \fB--fixlen\fP.
.TP
.NOP \f\*[B-Font]\-m\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-mtu\f[]=\f\*[I-Font]number\f[]
Override default MTU length (1500 bytes).
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  1 through MAX_SNAPLEN
.fi
.in -4
.sp
Override the default 1500 byte MTU size for determining the maximum padding length
(--fixlen=pad) or when truncating (--mtu-trunc).
.TP
.NOP \f\*[B-Font]\-\-mtu\-trunc\f[]
Truncate packets larger then specified MTU.
This option may appear up to 1 times.
.sp
Similar to \--fixlen, this option will truncate data in packets from Layer 3 and above to be
no larger then the MTU.
.TP
.NOP \f\*[B-Font]\-E\f[], \f\*[B-Font]\-\-efcs\f[]
Remove Ethernet checksums (FCS) from end of frames.
.sp
Note, this option is pretty dangerous!  We do not actually check to see if a FCS
actually exists in the frame, we just blindly delete the last 4 bytes.  Hence,
you should only use this if you know know that your OS provides the FCS when
reading raw packets.
.TP
.NOP \f\*[B-Font]\-\-ttl\f[]=\f\*[I-Font]string\f[]
Modify the IPv4/v6 TTL/Hop Limit.
.sp
Allows you to modify the TTL/Hop Limit of all the IPv4/v6 packets.  Specify a number to hard-code
the value or +/-value to increase or decrease by the value provided (limited to 1-255).
.sp
Examples:
.nf
    \--ttl=10
    \--ttl=+7
    \--ttl=-64
.fi
.TP
.NOP \f\*[B-Font]\-\-tos\f[]=\f\*[I-Font]number\f[]
Set the IPv4 TOS/DiffServ/ECN byte.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 255
.fi
.in -4
.sp
Allows you to override the TOS (also known as DiffServ/ECN) value in IPv4.
.TP
.NOP \f\*[B-Font]\-\-tclass\f[]=\f\*[I-Font]number\f[]
Set the IPv6 Traffic Class byte.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 255
.fi
.in -4
.sp
Allows you to override the IPv6 Traffic Class field.
.TP
.NOP \f\*[B-Font]\-\-flowlabel\f[]=\f\*[I-Font]number\f[]
Set the IPv6 Flow Label.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 1048575
.fi
.in -4
.sp
Allows you to override the 20bit IPv6 Flow Label field.  Has no effect on IPv4
packets.
.TP
.NOP \f\*[B-Font]\-F\f[] \f\*[I-Font]string\f[], \f\*[B-Font]\-\-fixlen\f[]=\f\*[I-Font]string\f[]
Pad or truncate packet data to match header length.
This option may appear up to 1 times.
.sp
Packets may be truncated during capture if the snaplen is smaller then the
packet.  This option allows you to modify the packet to pad the packet back
out to the size stored in the IPv4/v6 header or rewrite the IP header total length
to reflect the stored packet length.
.sp 1
\fBpad\fP
Truncated packets will be padded out so that the packet length matches the
IPv4 total length
.sp 1
\fBtrunc\fP
Truncated packets will have their IPv4 total length field rewritten to match
the actual packet length
.sp 1
\fBdel\fP
Delete the packet
.TP
.NOP \f\*[B-Font]\-\-fuzz\-seed\f[]=\f\*[I-Font]number\f[]
Fuzz 1 in X packets. Edit bytes, length, or emulate packet drop.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 0
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
This fuzzing was designed as to test layer 7 protocols such as voip protocols.
It modifies randomly 1 out of X packets (where X = \fB--fuzz-factor\fP) in order
for stateful protocols to cover more of their code.  The random fuzzing actions
focus on data start and end because it often is the part of the data application
protocols base their decisions on.
.sp
Possible fuzzing actions list:
 * drop packet
 * reduce packet size
 * edit packet Bytes:
   * Not all Bytes have the same probability of appearance in real life.
     Replace with 0x00, 0xFF, or a random byte with equal likelihood.
   * Not all Bytes have the same significance in a packet.
     Replace the start, the end, or the middle of the packet with equal likelihood.
 * do nothing (7 out of 8 packets)
.TP
.NOP \f\*[B-Font]\-\-fuzz\-factor\f[]=\f\*[I-Font]number\f[]
Set the Fuzz 1 in X packet ratio (default 1 in 8 packets).
This option must appear in combination with the following options:
fuzz-seed.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
greater than or equal to 1
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 8
.sp
Sets the ratio of for \fB--fuzz-seed\fP option. By default this value is 8,
which means 1 in 8 packets are modified by fuzzing. Note that this ratio is
based on the random number generated by the supplied fuzz seed. Therefore by
default you cannot expect that exactly every eighth packet will be modified.
.TP
.NOP \f\*[B-Font]\-\-skipl2broadcast\f[]
Skip rewriting broadcast/multicast Layer 2 addresses.
.sp
By default, editing Layer 2 addresses will rewrite 
broadcast and multicast MAC addresses.	Setting this flag
will keep broadcast/multicast MAC addresses from being rewritten.
.TP
.NOP \f\*[B-Font]\-\-dlt\f[]=\f\*[I-Font]string\f[]
Override output DLT encapsulation.
This option may appear up to 1 times.
.sp
By default, no DLT (data link type) conversion will be made.  
To change the DLT type of the output pcap, select one of the following values:
.sp 1
\fBenet\fP
Ethernet aka DLT_EN10MB
.sp 1
\fBhdlc\fP
Cisco HDLC aka DLT_C_HDLC
.sp 1
\fBjnpr_eth\fP
Juniper Ethernet DLT_C_JNPR_ETHER
.sp 1
\fBpppserial\fP
PPP Serial aka DLT_PPP_SERIAL
.sp 1
\fBuser\fP
User specified Layer 2 header and DLT type
.br
.TP
.NOP \f\*[B-Font]\-\-enet\-dmac\f[]=\f\*[I-Font]string\f[]
Override destination ethernet MAC addresses.
This option may appear up to 1 times.
.sp
Takes a pair of comma deliminated ethernet MAC addresses which
will replace the destination MAC address of outbound packets.
The first MAC address will be used for the server to client traffic
and the optional second MAC address will be used for the client
to server traffic.
.sp
Example:
.nf
    \--enet-dmac=00:12:13:14:15:16,00:22:33:44:55:66
.fi
.TP
.NOP \f\*[B-Font]\-\-enet\-smac\f[]=\f\*[I-Font]string\f[]
Override source ethernet MAC addresses.
This option may appear up to 1 times.
.sp
Takes a pair of comma deliminated ethernet MAC addresses which
will replace the source MAC address of outbound packets.
The first MAC address will be used for the server to client traffic
and the optional second MAC address will be used for the client 
to server traffic.
.sp
Example:
.nf
    \--enet-smac=00:12:13:14:15:16,00:22:33:44:55:66
.fi
.TP
.NOP \f\*[B-Font]\-\-enet\-subsmac\f[]=\f\*[I-Font]string\f[]
Substitute MAC addresses.
This option may appear up to 9999 times.
.sp
Allows you to rewrite ethernet MAC addresses of packets. It takes
comma delimited pair or MACs address and rewrites all occurrences of
the first MAC with the value of the second MAC.
Example:
.nf
    \--enet-subsmac=00:12:13:14:15:16,00:22:33:44:55:66
.fi
.TP
.NOP \f\*[B-Font]\-\-enet\-mac\-seed\f[]=\f\*[I-Font]number\f[]
Randomize MAC addresses.
This option may appear up to 1 times.
This option must not appear in combination with any of the following options:
enet-smac, enet-dmac, enet-subsmac.
This option takes an integer number as its argument.
.sp
Allows you to randomize ethernet MAC addresses of packets, mostly
like what \fB--seed\fP option does for IPv4/IPv6 addresses.
.TP
.NOP \f\*[B-Font]\-\-enet\-mac\-seed\-keep\-bytes\f[]=\f\*[I-Font]number\f[]
Randomize MAC addresses.
This option may appear up to 1 times.
This option must appear in combination with the following options:
enet-mac-seed.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  1 through 6
.fi
.in -4
.sp
Keep some bytes untouched when usinging \fB--enet-mac-seed\fP option.
.TP
.NOP \f\*[B-Font]\-\-enet\-vlan\f[]=\f\*[I-Font]string\f[]
Specify ethernet 802.1q VLAN tag mode.
This option may appear up to 1 times.
.sp
Allows you to rewrite ethernet frames to add a 802.1q header to standard 802.3
ethernet headers or remove the 802.1q VLAN tag information.
.sp 1
\fBadd\fP
Adds an 802.1q VLAN header to the existing 802.3 ethernet header. If
a VLAN header already exists, a new VLAN header is added outside of the
existing header.
.sp
Note that you will be allowed to run this option multiple times to create
more than 2 VLAN headers, however those packets will be valid. At most
you should have 2 X 802.1q VLAN tags, or outer an 802.1ad and an inner 802.1q
VLAN tag.
.sp 1
\fBdel\fP
Rewrites the existing 802.1q VLAN header as an 802.3 ethernet header
.TP
.NOP \f\*[B-Font]\-\-enet\-vlan\-tag\f[]=\f\*[I-Font]number\f[]
Specify the new ethernet 802.1q VLAN tag value.
This option may appear up to 1 times.
This option must appear in combination with the following options:
enet-vlan.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 4095
.fi
.in -4
.sp
.TP
.NOP \f\*[B-Font]\-\-enet\-vlan\-cfi\f[]=\f\*[I-Font]number\f[]
Specify the ethernet 802.1q VLAN CFI value.
This option may appear up to 1 times.
This option must appear in combination with the following options:
enet-vlan.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 1
.fi
.in -4
.sp
.TP
.NOP \f\*[B-Font]\-\-enet\-vlan\-pri\f[]=\f\*[I-Font]number\f[]
Specify the ethernet 802.1q VLAN priority.
This option may appear up to 1 times.
This option must appear in combination with the following options:
enet-vlan.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 7
.fi
.in -4
.sp
.TP
.NOP \f\*[B-Font]\-\-enet\-vlan\-proto\f[]=\f\*[I-Font]string\f[]
Specify VLAN tag protocol 802.1q or 802.1ad.
This option may appear up to 1 times.
.sp
Allows you to specify the protocol of the added VLAN tags.
.sp 1
\fB802.1q\fP
Specifies that 802.1q VLAN headers are to be added. This is the default.
.sp 1
\fB802.1ad\fP
Specifies that 802.1ad Q-in-Q VLAN headers are to be added. To make valid packets,
input packets must already have 802.1q VLAN headers.
.TP
.NOP \f\*[B-Font]\-\-hdlc\-control\f[]=\f\*[I-Font]number\f[]
Specify HDLC control value.
This option may appear up to 1 times.
This option takes an integer number as its argument.
.sp
The Cisco HDLC header has a 1 byte "control" field.  Apparently this should 
always be 0, but if you can use any 1 byte value.
.TP
.NOP \f\*[B-Font]\-\-hdlc\-address\f[]=\f\*[I-Font]number\f[]
Specify HDLC address.
This option may appear up to 1 times.
This option takes an integer number as its argument.
.sp
The Cisco HDLC header has a 1 byte "address" field which has two valid 
values:
.sp 1
\fB0x0F\fP
Unicast
.sp 1
\fB0xBF\fP
Broadcast
.br
You can however specify any single byte value.
.TP
.NOP \f\*[B-Font]\-\-user\-dlt\f[]=\f\*[I-Font]number\f[]
Set output file DLT type.
This option may appear up to 1 times.
This option takes an integer number as its argument.
.sp
Set the DLT value of the output pcap file.
.TP
.NOP \f\*[B-Font]\-\-user\-dlink\f[]=\f\*[I-Font]string\f[]
Rewrite Data-Link layer with user specified data.
This option may appear up to 2 times.
.sp
Provide a series of comma deliminated hex values which will be
used to rewrite or create the Layer 2 header of the packets.
The first instance of this argument will rewrite both server
and client traffic, but if this argument is specified a second
time, it will be used for the client traffic.
.sp
Example:
.nf
    \--user-dlink=01,02,03,04,05,06,00,1A,2B,3C,4D,5E,6F,08,00
.fi
.TP
.NOP \f\*[B-Font]\-d\f[] \f\*[I-Font]number\f[], \f\*[B-Font]\-\-dbug\f[]=\f\*[I-Font]number\f[]
Enable debugging output.
This option may appear up to 1 times.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]number\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 5
.fi
.in -4
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 0
.sp
If configured with \--enable-debug, then you can specify a verbosity 

release/src/router/tcpreplay/src/tcprewrite.c
/* $Id$ */

/*
 *   Copyright (c) 2001-2010 Aaron Turner <aturner at synfin dot net>
 *   Copyright (c) 2013-2022 Fred Klassen <tcpreplay at appneta dot com> - AppNeta
 *
 *   The Tcpreplay Suite of tools is free software: you can redistribute it
 *   and/or modify it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation, either version 3 of the
 *   License, or with the authors permission any later version.
 *
 *   The Tcpreplay Suite is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with the Tcpreplay Suite.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * Purpose: Modify packets in a pcap file based on rules provided by the
 * user to offload work from tcpreplay and provide an easier means of
 * reproducing traffic for testing purposes.
 */

#include "tcprewrite.h"
#include "config.h"
#include "common.h"
#include "tcpedit/fuzzing.h"
#include "tcpedit/tcpedit.h"
#include "tcprewrite_opts.h"
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#ifdef DEBUG
int debug;
#endif

#ifdef ENABLE_VERBOSE
/* tcpdump handle */
tcpdump_t tcpdump;
#endif

tcprewrite_opt_t options;
tcpedit_t *tcpedit;

/* local functions */
void tcprewrite_init(void);
void post_args(int argc, char *argv[]);
int rewrite_packets(tcpedit_t *tcpedit_ctx, pcap_t *pin, pcap_dumper_t *pout);

int
main(int argc, char *argv[])
{
    int optct, rcode;
    pcap_t *dlt_pcap;
#ifdef ENABLE_FRAGROUTE
    char ebuf[FRAGROUTE_ERRBUF_LEN];
#endif
    tcprewrite_init();

    /* call autoopts to process arguments */
    optct = optionProcess(&tcprewriteOptions, argc, argv);
    argc -= optct;
    argv += optct;

    /* parse the tcprewrite args */
    post_args(argc, argv);

    /* init tcpedit context */
    if (tcpedit_init(&tcpedit, pcap_datalink(options.pin)) < 0) {
        err_no_exitx("Error initializing tcpedit: %s", tcpedit_geterr(tcpedit));
        tcpedit_close(&tcpedit);
        exit(-1);
    }

    /* parse the tcpedit args */
    rcode = tcpedit_post_args(tcpedit);
    if (rcode < 0) {
        err_no_exitx("Unable to parse args: %s", tcpedit_geterr(tcpedit));
        tcpedit_close(&tcpedit);
        exit(-1);
    } else if (rcode == 1) {
        warnx("%s", tcpedit_geterr(tcpedit));
    }

    if (tcpedit_validate(tcpedit) < 0) {
        err_no_exitx("Unable to edit packets given options:\n%s", tcpedit_geterr(tcpedit));
        tcpedit_close(&tcpedit);
        exit(-1);
    }

    /* fuzzing init */
    fuzzing_init(tcpedit->fuzz_seed, tcpedit->fuzz_factor);

    /* open up the output file */
    options.outfile = safe_strdup(OPT_ARG(OUTFILE));
    dbgx(1, "Rewriting DLT to %s", pcap_datalink_val_to_name(tcpedit_get_output_dlt(tcpedit)));
    if ((dlt_pcap = pcap_open_dead(tcpedit_get_output_dlt(tcpedit), 65535)) == NULL) {
        tcpedit_close(&tcpedit);
        err(-1, "Unable to open dead pcap handle.");
    }

    dbgx(1, "DLT of dlt_pcap is %s", pcap_datalink_val_to_name(pcap_datalink(dlt_pcap)));

#ifdef ENABLE_FRAGROUTE
    if (options.fragroute_args) {
        if ((options.frag_ctx = fragroute_init(65535, pcap_datalink(dlt_pcap), options.fragroute_args, ebuf)) == NULL) {
            err_no_exitx("%s", ebuf);
            tcpedit_close(&tcpedit);
            exit(-1);
        }
    }
#endif

#ifdef ENABLE_VERBOSE
    if (options.verbose) {
        tcpdump_open(&tcpdump, dlt_pcap);
    }
#endif

    if ((options.pout = pcap_dump_open(dlt_pcap, options.outfile)) == NULL) {
        err_no_exitx("Unable to open output pcap file: %s", pcap_geterr(dlt_pcap));
        tcpedit_close(&tcpedit);
        exit(-1);
    }

    pcap_close(dlt_pcap);

    /* rewrite packets */
    if (rewrite_packets(tcpedit, options.pin, options.pout) == TCPEDIT_ERROR) {
        err_no_exitx("Error rewriting packets: %s", tcpedit_geterr(tcpedit));
        tcpedit_close(&tcpedit);
        exit(-1);
    }

    /* clean up after ourselves */
    pcap_dump_close(options.pout);
    pcap_close(options.pin);
    tcpedit_close(&tcpedit);

#ifdef ENABLE_VERBOSE
    tcpdump_close(&tcpdump);
#endif

#ifdef ENABLE_FRAGROUTE
    if (options.frag_ctx) {
        fragroute_close(options.frag_ctx);
    }
#endif

#ifdef ENABLE_DMALLOC
    dmalloc_shutdown();
#endif

    restore_stdin();
    return 0;
}

void
tcprewrite_init(void)
{
    memset(&options, 0, sizeof(options));

#ifdef ENABLE_VERBOSE
    /* clear out tcpdump struct */
    memset(&tcpdump, '\0', sizeof(tcpdump_t));
#endif

    if (fcntl(STDERR_FILENO, F_SETFL, O_NONBLOCK) < 0)
        warnx("Unable to set STDERR to non-blocking: %s", strerror(errno));
}

/**
 * post AutoGen argument processing
 */
void
post_args(_U_ int argc, _U_ char *argv[])
{
    char ebuf[PCAP_ERRBUF_SIZE];

#ifdef DEBUG
    if (HAVE_OPT(DBUG))
        debug = OPT_VALUE_DBUG;
#else
    if (HAVE_OPT(DBUG))
        warn("not configured with --enable-debug.  Debugging disabled.");
#endif

#ifdef ENABLE_VERBOSE
    if (HAVE_OPT(VERBOSE))
        options.verbose = 1;

    if (HAVE_OPT(DECODE))
        tcpdump.args = safe_strdup(OPT_ARG(DECODE));
#endif

#ifdef ENABLE_FRAGROUTE
    if (HAVE_OPT(FRAGROUTE))
        options.fragroute_args = safe_strdup(OPT_ARG(FRAGROUTE));

    options.fragroute_dir = FRAGROUTE_DIR_BOTH;
    if (HAVE_OPT(FRAGDIR)) {
        if (strcmp(OPT_ARG(FRAGDIR), "c2s") == 0) {
            options.fragroute_dir = FRAGROUTE_DIR_C2S;
        } else if (strcmp(OPT_ARG(FRAGDIR), "s2c") == 0) {
            options.fragroute_dir = FRAGROUTE_DIR_S2C;
        } else if (strcmp(OPT_ARG(FRAGDIR), "both") == 0) {
            options.fragroute_dir = FRAGROUTE_DIR_BOTH;
        } else {
            errx(-1, "Unknown --fragdir value: %s", OPT_ARG(FRAGDIR));
        }
    }
#endif

    /* open up the input file */
    options.infile = safe_strdup(OPT_ARG(INFILE));
    if ((options.pin = pcap_open_offline(options.infile, ebuf)) == NULL)
        errx(-1, "Unable to open input pcap file: %s", ebuf);

#ifdef HAVE_PCAP_SNAPSHOT
    if (pcap_snapshot(options.pin) < 65535)
        warnx("%s was captured using a snaplen of %d bytes.  This may mean you have truncated packets.",
              options.infile,
              pcap_snapshot(options.pin));
#endif
}

/**
 * Main loop to rewrite packets
 */
int
rewrite_packets(tcpedit_t *tcpedit_ctx, pcap_t *pin, pcap_dumper_t *pout)
{
    tcpr_dir_t cache_result = TCPR_DIR_C2S; /* default to primary */
    struct pcap_pkthdr pkthdr, *pkthdr_ptr; /* packet header */
    const u_char *pktconst = NULL;          /* packet from libpcap */
    u_char **pktdata = NULL;
    static u_char *pktdata_buff;
    static char *frag = NULL;
    COUNTER packetnum = 0;
    int rcode;
#ifdef ENABLE_FRAGROUTE
    int frag_len, proto;
#endif

    pkthdr_ptr = &pkthdr;

    if (pktdata_buff == NULL)
        pktdata_buff = (u_char *)safe_malloc(MAXPACKET);

    pktdata = &pktdata_buff;

    if (frag == NULL)
        frag = (char *)safe_malloc(MAXPACKET);

    /* MAIN LOOP
     * Keep sending while we have packets or until
     * we've sent enough packets
     */
    while ((pktconst = safe_pcap_next(pin, pkthdr_ptr)) != NULL) {
        packetnum++;
        dbgx(2, "packet " COUNTER_SPEC " caplen %d", packetnum, pkthdr.caplen);

        if (pkthdr.caplen > MAX_SNAPLEN)
            errx(-1, "Frame too big, caplen %d exceeds %d", pkthdr.caplen, MAX_SNAPLEN);
        /*
         * copy over the packet so we can pad it out if necessary and
         * because pcap_next() returns a const ptr
         */
        memcpy(*pktdata, pktconst, pkthdr.caplen);

        /* Dual nic processing? */
        if (options.cachedata != NULL) {
            cache_result = check_cache(options.cachedata, packetnum);
        }

        /* sometimes we should not send the packet, in such cases
         * no point in editing this packet at all, just write it to the
         * output file (note, we can't just remove it, or the tcpprep cache
         * file will lose it's indexing
         */

        if (cache_result == TCPR_DIR_NOSEND)
            goto WRITE_PACKET; /* still need to write it so cache stays in sync */

        if ((rcode = tcpedit_packet(tcpedit_ctx, &pkthdr_ptr, pktdata, cache_result)) == TCPEDIT_ERROR) {
            return rcode;
        } else if ((rcode == TCPEDIT_SOFT_ERROR) && HAVE_OPT(SKIP_SOFT_ERRORS)) {
            /* don't write packet */
            dbgx(1, "Packet " COUNTER_SPEC " is suppressed from being written due to soft errors", packetnum);
            continue;
        }

#ifdef ENABLE_VERBOSE
        if (options.verbose)
            tcpdump_print(&tcpdump, pkthdr_ptr, *pktdata);
#endif

WRITE_PACKET:
#ifdef ENABLE_FRAGROUTE
        if (options.frag_ctx == NULL) {
            /* write the packet when there's no fragrouting to be done */
            if (pkthdr_ptr->caplen)
                pcap_dump((u_char *)pout, pkthdr_ptr, *pktdata);
        } else {
            /* get the L3 protocol of the packet */
            proto = tcpedit_l3proto(tcpedit_ctx, AFTER_PROCESS, *pktdata, pkthdr_ptr->caplen);

            /* packet is IPv4/IPv6 AND needs to be fragmented */
            if ((proto == ETHERTYPE_IP || proto == ETHERTYPE_IP6) &&
                ((options.fragroute_dir == FRAGROUTE_DIR_BOTH) ||
                 (cache_result == TCPR_DIR_C2S && options.fragroute_dir == FRAGROUTE_DIR_C2S) ||
                 (cache_result == TCPR_DIR_S2C && options.fragroute_dir == FRAGROUTE_DIR_S2C))) {
#ifdef DEBUG
                int i = 0;
#endif
                if (fragroute_process(options.frag_ctx, *pktdata, pkthdr_ptr->caplen) < 0)
                    errx(-1, "Error processing packet via fragroute: %s", options.frag_ctx->errbuf);

                while ((frag_len = fragroute_getfragment(options.frag_ctx, &frag)) > 0) {
                    /* frags get the same timestamp as the original packet */
                    dbgx(1, "processing packet " COUNTER_SPEC " frag: %u (%d)", packetnum, i++, frag_len);
                    pkthdr_ptr->caplen = frag_len;
                    pkthdr_ptr->len = frag_len;
                    if (pkthdr_ptr->caplen)
                        pcap_dump((u_char *)pout, pkthdr_ptr, (u_char *)frag);
                }
            } else {
                /* write the packet without fragroute */
                if (pkthdr_ptr->caplen)
                    pcap_dump((u_char *)pout, pkthdr_ptr, *pktdata);
            }
        }
#else
        /* write the packet when there's no fragrouting to be done */
        if (pkthdr_ptr->caplen)
            pcap_dump((u_char *)pout, pkthdr_ptr, *pktdata);

#endif
    } /* while() */
    return 0;
}
